drafts/homemade-signals.md

```diff
@@ -0,0 +1,232 @@
+
+In [a previous blog post](https://software-engineering-corner.zuehlke.com/signals-signals-everywhere), we explored where Signals came from, what they are, and how they took over the frontend ecosystem.
+We already saw that they simplify keeping the UI in sync with the state straightforwardly, even if we skip using any other libraries.
+But today, we're taking it one step further.
+We are building rudimentary Signals from scratch to understand the underlying mechanisms.
+This exercise has been massively valuable for me.
+
+## Quick Recap
+
+Let's do a quick recap.
+Signals have taken over the world of frontend frameworks.
+Just about every framework except React has hopped on board with this concept over the past two years, and it looks like they're to stay.
+
+JavaScript frameworks try to solve the challenge of keeping the UI consistent and up-to-date with the state.
+After exploring a wide range of solutions, it seems that most agree that Signals are an essential piece of the puzzle.
+They are special JS objects that can notify subscribers about changes.
+We can pass around a Signal without it losing reactivity.
+There are three types of reactivity: reactive state, derived values, and side effects.
+
+Effects are the powerhouses that make everything move, as they produce a side effect every time a source Signal changes.
+That allows frameworks to re-render relevant segments of the UI much more efficiently than elaborate change detectors or by comparing virtual DOM trees.
+
+And now, off to building Signals ourselves.
+
+## Basic Reactive State
+
+The first element we're implementing is the elementary building block: a Signal.
+This initial version is just a simple wrapper around a value with accessor functions.
+
+```js
+function signal(initialValue) {
+    let _value = initialValue;
+    return {
+        get value() {
+            return _value;
+        },
+        set value(newValue) {
+            if (newValue !== _value) {
+                _value = newValue;
+            }
+        }
+    }
+}
+```
+
+It doesn't notify anyone reading the value yet.
+But we can pass its reference around and always get the up-to-date value whenever we read it.
+That is so far only useful for primitive values since JavaScript otherwise passes them by value.
+
+```js
+const counter = signal(0);
+counter.value += 1;
+console.log(counter.value); // 1
+```
+
+## Derived Values
+
+Signals that are derived, or _computed_, from other Signals are an importation use case.
+We want to be able to use them like this:
+
+```js
+const counter = signal(0);
+const double = computed(()= => counter.value * 2);
+```
+
+Due to their calculated nature, we only need them to be readable, not writable.
+We want it to re-run the given computation every time we read it.
+A very simple version could look something like the following:
+
+```js
+function computed(computation) {
```

Comment:
does computed also need something special like effect, or is it just always updating correctly?

Reply:
just works, I added a quick sentence on line 160